supplementary/alica_templates/Expr/src/ActionServerExample2379894799421542548.cpp

```diff
@@ -0,0 +1,137 @@
+ *   - Comment:
+ *   - Source2Dest: WaitForGoal --> ExecuteGoal
+ *
+ * Precondition: 1886820548377048134 (1886820548377048134)
+ *   - Enabled: true
+ *   - PluginName: DefaultPlugin
+ *   - ConditionString:
+ *   - Variables:
+ *   - Quantifiers:
+ *
+ * Abstract Plans in WaitForGoal:
+ */
+bool PreCondition1886820548377048134::evaluate(std::shared_ptr<RunningPlan> rp, const IAlicaWorldModel* wm)
+{
+    /*PROTECTED REGION ID(430744406068167347) ENABLED START*/
+    LockedBlackboardRO bb = LockedBlackboardRO(*(rp->getBasicPlan()->getBlackboard()));
+    return bb.get<std::optional<int32_t>>("goal").has_value();
+    /*PROTECTED REGION END*/
+}
+
+/**
+ * Transition: 1354699620997961969 (1354699620997961969)
+ *   - Comment:
+ *   - Source2Dest: ExecuteGoal --> WaitForGoal
+ *
+ * Precondition: 587249152722263568 (587249152722263568)
+ *   - Enabled: true
+ *   - PluginName: DefaultPlugin
+ *   - ConditionString:
+ *   - Variables:
+ *   - Quantifiers:
+ *
+ * Abstract Plans in ExecuteGoal:
+ *   - DummyImplementation (4126421719858579722)
+ */
+bool PreCondition587249152722263568::evaluate(std::shared_ptr<RunningPlan> rp, const IAlicaWorldModel* wm)
+{
+    /*PROTECTED REGION ID(1354699620997961969) ENABLED START*/
+    LockedBlackboardRO bb = LockedBlackboardRO(*(rp->getBasicPlan()->getBlackboard()));
+    return rp->isAnyChildTaskSuccessful() || rp->isAnyChildStatus(PlanStatus::Failed) || bb.get<std::optional<bool>>("cancel").value_or(false);
+    /*PROTECTED REGION END*/
+}
+
+/*PROTECTED REGION ID(methods2379894799421542548) ENABLED START*/
+// Add additional options here
+void ActionServerExample2379894799421542548::onInit()
+{
+    LockedBlackboardRW bb = LockedBlackboardRW(*getBlackboard());
+    bb.registerValue("result", std::optional<int32_t>());
+    bb.registerValue("feedback", std::optional<int32_t>());
+    bb.registerValue("goal", std::optional<int32_t>());
+    bb.registerValue("cancel", std::optional<bool>());
+}
+
+void ActionServerExample2379894799421542548::run(void* msg)
+{
+    LockedBlackboardRW bb = LockedBlackboardRW(*getBlackboard());
+    if (_actionServer->isNewGoalAvailable()) {
+        bb.get<std::optional<int32_t>>("cancel") = false;
+        // on the arrival of a new goal, preempt the current goal
+        if (_actionServer->isActive()) {
+            _actionServer->setPreempted();
+        }
+        // dont send result of the old goal when a new goal is available
+        bb.get<std::optional<int32_t>>("result") = std::nullopt;
+        bb.get<std::optional<int32_t>>("goal") = _actionServer->acceptNewGoal()->value;
+    }
+    if (_actionServer->isPreemptRequested()) {
+        _actionServer->setAborted();
+        // dont send result of the cancelled goal, clear cancelled goal
+        bb.get<std::optional<int32_t>>("result") = std::nullopt;
+        bb.get<std::optional<int32_t>>("goal") = std::nullopt;
+        bb.get<std::optional<int32_t>>("cancel") = true;
```

Comment:
For pre-emption, we don't want "cancel" to be true as this will trigger currently running DummyImplementation to be terminated. We want to let the action server plan decide how it wants to handle the preemption and new goal.

Reply:
I think isPreemptRequested is actually called when the action client cancels the goal. The name is misleading. If a new goal is requested by the client without cancelling the current goal (i.e. preemption case) then isNewGoalAvailable will be true & the new goal will be accepted by the above if case.